File: solutions/go/food-chain/1/food_chain.go
package foodchain

import (
    "fmt"
    "strings"
)

var creatures = []string{"fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse",
                        "spider that wriggled and jiggled and tickled inside her"}

var comments = []string{"I know an old lady who swallowed a %s.",
                     "She swallowed the %s to catch the %s.","",
                     "I don't know why she swallowed the fly. Perhaps she'll die.",
                     "It wriggled and jiggled and tickled inside her.",
                     "How absurd to swallow a bird!",
                     "Imagine that, to swallow a cat!",
                     "What a hog, to swallow a dog!",
                     "Just opened her throat and swallowed a goat!",
                     "I don't know how she swallowed a cow!",
                     "She's dead, of course!"}

func Verse(v int) string {
	res := []string{fmt.Sprintf(comments[0], creatures[v-1]), comments[v+2]}
    if v > 1 && v < 8 {
        for a := v -1; a >= 1; a -= 1 {
            k := a-1; if a == 2 {k = 8}
            verse := fmt.Sprintf(comments[1], creatures[a], creatures[k])
            res = append(res, verse)
        }
        res = append(res, comments[3])
    }
    return strings.Join(res, "\n")
}

func Verses(start, end int) string {
	res := []string{}
    for v := start; v <= end; v += 1 {
        res = append(res, Verse(v))
    }
    return strings.Join(res, "\n\n")
}

func Song() string {
	return Verses(1, 8)
}

File: solutions/ruby/food-chain/1/food_chain.rb
class FoodChain
  VERSE = [
    { name: 'fly', phrase: "I don't know why she swallowed the fly. Perhaps she'll die." },
    { name: 'spider', phrase: 'It wriggled and jiggled and tickled inside her.', add: ' that wriggled and jiggled and tickled inside her'},
    { name: 'bird', phrase: 'How absurd to swallow a bird!' },
    { name: 'cat', phrase: 'Imagine that, to swallow a cat!' },
    { name: 'dog', phrase: 'What a hog, to swallow a dog!' },
    { name: 'goat', phrase: 'Just opened her throat and swallowed a goat!' },
    { name: 'cow', phrase: "I don't know how she swallowed a cow!" },
    { name: 'horse', phrase: "She's dead, of course!"}
  ]
  def self.song
    (0...VERSE.size).map { |n| verse(n, '') }.join("\n")
  end
  def self.verse(n, acc)
    acc << "I know an old lady who swallowed a %{name}.\n%{phrase}\n" % { name: VERSE[n][:name], phrase: VERSE[n][:phrase] } if acc == ''
    return acc if n == 0 || VERSE[n][:name] == 'horse'
    acc << "She swallowed the %{name} to catch the %{name_prev}%{add}.\n" % { name: VERSE[n][:name], name_prev: VERSE[n-1][:name], add: VERSE[n-1][:add] || '' }
    acc << (VERSE[n- 1][:phrase] + "\n") if n == 1
    verse(n - 1, acc)
  end
end
